File: 329-LongestIncreasingPathInAMatrix/Solution.java
/**
 * 329. Longest Increasing Path in a Matrix
 * Tag: DFS, DP
 * Chang Li at UC San Diego
 * Jan. 21, 2016
 */

public class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }
        int[][] table = new int[matrix.length][matrix[0].length];
        int maxLen = 0;
        for(int i=0; i<matrix.length; i++) {
            for(int j=0; j<matrix[0].length; j++) {
                maxLen = Math.max(maxLen, DFS(matrix, i, j, table));
            }
        }
        return maxLen;
    }
    
    public int DFS(int[][] matrix, int x, int y, int[][] table) {
        if(table[x][y] != 0) {
            return table[x][y];
        }
        int maxLen = 1;
        if(x != 0 && matrix[x-1][y] > matrix[x][y]) {
            maxLen = Math.max(maxLen, 1+DFS(matrix, x-1, y, table));
        }
        if(x != matrix.length-1 && matrix[x+1][y] > matrix[x][y]) {
            maxLen = Math.max(maxLen, 1+DFS(matrix, x+1, y, table));
        }
        if(y != 0 && matrix[x][y-1] > matrix[x][y]) {
            maxLen = Math.max(maxLen, 1+DFS(matrix, x, y-1, table));
        }
        if(y != matrix[0].length-1 && matrix[x][y+1] > matrix[x][y]) {
            maxLen = Math.max(maxLen, 1+DFS(matrix, x, y+1, table));
        }
        table[x][y] = maxLen;
        return maxLen;
    }
}

File: 329-LongestIncreasingPathInAMatrix/solution.py
class Solution:
    def get_adj_pos(self, i: int, j: int, mat: List[List[int]]) -> List[List[int]]:
      res = []
      for di, dj in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
        new_i, new_j = i + di, j + dj
        if (0 <= new_i < len(mat)) and (0 <= new_j < len(mat[0])):
          res.append((new_i, new_j))
      return res
   
    def _helper(self, i: int, j: int, states: List[List[int]], matrix: List[List[int]]) -> int:
      if states[i][j] != 0:
        return states[i][j]
      longest = 0
      for new_i, new_j in self.get_adj_pos(i, j, matrix):
        if matrix[new_i][new_j] > matrix[i][j]:
          longest = max(longest, self._helper(new_i, new_j, states, matrix))
      states[i][j] = longest + 1
      return states[i][j]

    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
      states = [[0] * len(matrix[0]) for _ in range(len(matrix))]
      res = 0
      for i in range(len(matrix)):
        for j in range(len(matrix[0])):
          res = max(res, self._helper(i, j, states, matrix))

      # for i in range(len(matrix)):
      #   print(states[i])

      return res
